lib/membrane_dashboard/dagre.ex

```diff
@@ -61,20 +61,51 @@ defmodule Membrane.Dashboard.Dagre do
 
   @spec query(non_neg_integer(), non_neg_integer()) :: {:ok, any()} | {:error, any()}
   def query(time_from, time_to) do
-    result =
-      """
-      SELECT parent_path, l.from, l.to, pad_from, pad_to FROM links l WHERE time BETWEEN \
-      '#{parse_time(time_from)}' AND '#{parse_time(time_to)}'
-      """
-      |> Repo.query()
+    import Ecto.Query
+
+    query_links = fn elements ->
+      from(l in "links",
+        where:
+          fragment(
+            "ARRAY[?, ?, ?, ?] && ?",
+            fragment("concat(?, '/', ?)", l.parent_path, l.from),
+            fragment("concat(?, '/', ?)", l.parent_path, l.to),
+            fragment("concat(?, '/', ?, ' bin')", l.parent_path, l.from),
+            fragment("concat(?, '/', ?, ' bin')", l.parent_path, l.to),
+            fragment("?", ^elements)
+          ),
+        select: [l.parent_path, l.from, l.to, l.pad_from, l.pad_to]
+      )
+      |> Repo.all()
+    end
 
-    with {:ok, %Postgrex.Result{rows: links}} <- result,
-         {:ok, dagre} <- links |> format_rows() |> __MODULE__.G6Marshaller.run() do
+    with {:ok, all_alive_elements} <- alive_element_paths(time_from, time_to),
+         {:ok, dead_elements} <- element_paths_in_time_range(time_from, time_to, true),
+         {:ok, new_elements} <- element_paths_in_time_range(time_from, time_to, false),
+         existing_elements <- MapSet.difference(all_alive_elements, new_elements),
+         elements_to_query <- MapSet.union(all_alive_elements, dead_elements),
+         links <- elements_to_query |> Enum.map(&String.replace(&1, "\\", "")) |> query_links.(),
+         {:ok, dagre} <-
+           links
+           |> format_rows()
+           |> __MODULE__.G6Marshaller.run(
+             dead: dead_elements,
+             new: new_elements,
+             existing: existing_elements
+           ) do
       %{nodes: nodes, edges: edges, combos: combos} = dagre
-      [nodes, edges, combos] = [nodes, edges, combos] |> Enum.map(&MapSet.to_list/1)
+
+      [nodes, edges, combos] =
+        [nodes, edges, combos]
+        |> Enum.map(fn
+          %MapSet{} = set -> MapSet.to_list(set)
+          list -> list
+        end)
+
       {:ok, %{nodes: nodes, edges: edges, combos: combos}}
     else
-      {:error, _reason} -> {:error, "Failed to fetch links"}
+      {:error, reason} ->
+        {:error, reason}
```

Comment:
I think there is no need of `else` if we are not changing anything in the error

Reply:
Good catch, although (not in this case) sometimes is helpful to catch unmatched things that are not errors as we could get pattern match error.